File: example-bots/python-portfolio-tracker/query.py
"""
Portfolio Tracker Query Handlers
================================
Query endpoints for the portfolio tracker bot.

These handlers provide read-only access to bot state and computed data.
They can be executed via:
  - CLI: the0 bot query <bot_id> /portfolio
  - API: POST /bot/:id/query

Available queries:
  /portfolio - Get current portfolio summary with history
  /status    - Get bot configuration and state summary
  /history   - Get portfolio value history with optional limit
  /trades    - Get trade statistics
"""

from the0 import query, state


@query.handler("/portfolio")
def get_portfolio(req):
    """
    Get current portfolio summary.

    Returns:
        positions: Number of history entries
        total_trades: Total trades executed
        latest_value: Most recent portfolio value
        history: Last 10 portfolio values
    """
    portfolio_history = state.get("portfolio_history", [])
    total_trades = state.get("total_trades", 0)

    latest_value = None
    if portfolio_history:
        latest_value = portfolio_history[-1].get("value")

    return {
        "history_count": len(portfolio_history),
        "total_trades": total_trades,
        "latest_value": latest_value,
        "recent_history": portfolio_history[-10:] if portfolio_history else [],
    }


@query.handler("/status")
def get_status(req):
    """
    Get bot status and configuration.

    Returns:
        config: Bot configuration
        state_summary: Summary of persisted state
    """
    config = query.get_config()
    portfolio_history = state.get("portfolio_history", [])
    total_trades = state.get("total_trades", 0)

    return {
        "config": {
            "initial_value": config.get("initial_value", 10000),
            "volatility": config.get("volatility", 0.02),
            "symbols": config.get("symbols", ["BTC", "ETH", "SOL"]),
        },
        "state_summary": {
            "history_entries": len(portfolio_history),
            "total_trades": total_trades,
            "has_data": len(portfolio_history) > 0,
        },
    }


@query.handler("/history")
def get_history(req):
    """
    Get portfolio value history.

    Query params:
        limit: Maximum number of entries to return (default: 50)

    Returns:
        history: List of {value, timestamp} entries
        count: Total number of entries returned
        total: Total entries in history
    """
    limit = int(req.get("limit", 50))
    portfolio_history = state.get("portfolio_history", [])

    # Return most recent entries up to limit
    history = portfolio_history[-limit:] if portfolio_history else []

    return {
        "history": history,
        "count": len(history),
        "total": len(portfolio_history),
    }


@query.handler("/trades")
def get_trades(req):
    """
    Get trade statistics.

    Returns:
        total_trades: Number of trades executed
        average_per_run: Average trades per run (estimated)
    """
    portfolio_history = state.get("portfolio_history", [])
    total_trades = state.get("total_trades", 0)

    runs = len(portfolio_history)
    avg_per_run = total_trades / runs if runs > 0 else 0

    return {
        "total_trades": total_trades,
        "total_runs": runs,
        "average_per_run": round(avg_per_run, 2),
    }


if __name__ == "__main__":
    query.run()

File: example-bots/typescript-price-alerts/query.ts
/**
 * Price Alerts Query Handlers
 * ===========================
 * Query endpoints for the price alerts bot.
 *
 * These handlers provide read-only access to bot state and computed data.
 * They can be executed via:
 *   - CLI: the0 bot query <bot_id> /current-price
 *   - API: POST /bot/:id/query
 *
 * Available queries:
 *   /current-price - Get latest price with 24h range and change
 *   /alerts        - Get alert count and statistics
 *   /statistics    - Get overall bot statistics
 */

import { query, state } from "@alexanderwanyoike/the0-node";

// Persistent state structure (matches main.ts)
interface PersistedState {
  priceHistory: number[];
  alertCount: number;
}

/**
 * Get current price information.
 *
 * Returns:
 *   current_price: Latest price value
 *   high_24h: Highest price in history
 *   low_24h: Lowest price in history
 *   change_pct: Percentage change from oldest to newest
 *   data_points: Number of price data points
 */
query.handler("/current-price", (req) => {
  const persisted = state.get<PersistedState>("bot_state", {
    priceHistory: [],
    alertCount: 0,
  });

  const history = persisted.priceHistory;

  if (history.length === 0) {
    return {
      status: "no_data",
      message: "No price history available",
    };
  }

  const currentPrice = history[history.length - 1];
  const high24h = Math.max(...history);
  const low24h = Math.min(...history);
  const changePct =
    history.length > 1
      ? ((currentPrice - history[0]) / history[0]) * 100
      : 0;

  return {
    current_price: Math.round(currentPrice * 100) / 100,
    high_24h: Math.round(high24h * 100) / 100,
    low_24h: Math.round(low24h * 100) / 100,
    change_pct: Math.round(changePct * 1000) / 1000,
    data_points: history.length,
  };
});

/**
 * Get alert information.
 *
 * Query params:
 *   limit: Maximum number of alerts to return (for future use)
 *
 * Returns:
 *   total_alerts: Total alerts generated
 *   price_data_points: Number of price history entries
 */
query.handler("/alerts", (req) => {
  const limit = parseInt(req.get("limit", "10") || "10", 10);
  const persisted = state.get<PersistedState>("bot_state", {
    priceHistory: [],
    alertCount: 0,
  });

  return {
    total_alerts: persisted.alertCount,
    limit_applied: limit,
    // Note: Individual alerts are not persisted, only the count
    // To return actual alerts, we would need to persist alert history
  };
});

/**
 * Get bot statistics.
 *
 * Returns:
 *   total_alerts: Total alerts generated
 *   price_data_points: Number of price entries
 *   config: Bot configuration summary
 */
query.handler("/statistics", (req) => {
  const config = query.getConfig();
  const persisted = state.get<PersistedState>("bot_state", {
    priceHistory: [],
    alertCount: 0,
  });

  return {
    total_alerts: persisted.alertCount,
    price_data_points: persisted.priceHistory.length,
    config: {
      symbol: config.symbol || "BTC/USD",
      alert_threshold: config.alert_threshold || 1.0,
      update_interval_ms: config.update_interval_ms || 5000,
    },
  };
});

// Run query system
query.run();
